File: internal/edgeworkers/project.go
package edgeworkers

import (
	"encoding/json/jsontext"
	json "encoding/json/v2"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"golang.org/x/text/cases"
	"golang.org/x/text/collate"
	"golang.org/x/text/language"
)

const ProjectFile = "edge-workers.json"
const ManifestFile = "worker.json"
const WorkersDir = "workers"
const BuildDir = "build"
const ConventionalDir = "edge-workers"

type ProjectDescriptor struct {
	Type string
	SDK  *string
}
type Manifest struct {
	Name, Entry string
	Location    LocationValue
	OnFailure   *string
}
type LocalWorker struct {
	Dir      string
	Manifest Manifest
}

func pathExists(path string) bool { _, err := os.Stat(path); return err == nil }

func ResolveProjectDir(cwd string, explicit *string) (string, error) {
	cwd, err := filepath.Abs(cwd)
	if err != nil {
		return "", err
	}
	if explicit != nil {
		if *explicit == "" {
			return "", errors.New("The --path flag requires a path to the edge-workers project.")
		}
		target := *explicit
		if !filepath.IsAbs(target) {
			target = filepath.Join(cwd, target)
		}
		target = filepath.Clean(target)
		if !pathExists(filepath.Join(target, ProjectFile)) {
			return "", fmt.Errorf("No edge-workers project found at \"%s\" (missing %s).", target, ProjectFile)
		}
		return target, nil
	}
	for current := cwd; ; current = filepath.Dir(current) {
		if pathExists(filepath.Join(current, ProjectFile)) {
			return current, nil
		}
		if filepath.Dir(current) == current {
			break
		}
	}
	conventional := filepath.Join(cwd, ConventionalDir)
	if pathExists(filepath.Join(conventional, ProjectFile)) {
		return conventional, nil
	}
	return "", errors.New("No edge-workers project found here. Run `vip edge-workers init` to create one, run the command from inside a project, or pass `--path` to point at one.")
}

func readProjectJSON(file, label string) (any, error) {
	info, err := os.Lstat(file)
	if err != nil {
		return nil, fmt.Errorf("Could not read %s at \"%s\".", label, file)
	}
	if info.Mode()&os.ModeSymlink != 0 {
		return nil, fmt.Errorf("%s at \"%s\" must not be a symbolic link.", label, file)
	}
	data, err := os.ReadFile(file)
	if err != nil {
		return nil, fmt.Errorf("Could not read %s at \"%s\".", label, file)
	}
	var value any
	// JSON.parse accepts duplicate properties (last wins) and replaces invalid UTF-8.
	err = json.Unmarshal(data, &value, jsontext.AllowDuplicateNames(true), jsontext.AllowInvalidUTF8(true))
	if err != nil {
		return nil, fmt.Errorf("%s at \"%s\" is not valid JSON.", strings.ToUpper(label[:1])+label[1:], file)
	}
	return value, nil
}

func ReadProjectDescriptor(projectDir string) (ProjectDescriptor, error) {
	file := filepath.Join(projectDir, ProjectFile)
	value, err := readProjectJSON(file, "project descriptor")
	if err != nil {
		return ProjectDescriptor{}, err
	}
	m, ok := value.(map[string]any)
	if !ok {
		return ProjectDescriptor{}, fmt.Errorf("Project descriptor at \"%s\" has an invalid \"type\" field.", file)
	}
	if _, present := m["type"]; !present {
		return ProjectDescriptor{}, fmt.Errorf("Project descriptor at \"%s\" is missing a \"type\" field.", file)
	}
	if m["type"] != "assemblyscript" {
		return ProjectDescriptor{}, fmt.Errorf("Project descriptor at \"%s\" has an invalid \"type\" field.", file)
	}
	d := ProjectDescriptor{Type: "assemblyscript"}
	if raw, present := m["sdk"]; present {
		sdk, ok := raw.(string)
		if !ok {
			return d, fmt.Errorf("Project descriptor at \"%s\" has an invalid \"sdk\" field.", file)
		}
		d.SDK = &sdk
	}
	return d, nil
}

func ReadManifest(workerDir string) (Manifest, error) {
	file := filepath.Join(workerDir, ManifestFile)
	value, err := readProjectJSON(file, "worker manifest")
	if err != nil {
		return Manifest{}, err
	}
	m, ok := value.(map[string]any)
	if !ok {
		return Manifest{}, fmt.Errorf("Worker manifest at \"%s\" must be an object.", file)
	}
	name, ok := m["name"].(string)
	if !ok {
		raw, present := m["name"]
		return Manifest{}, fmt.Errorf("Invalid worker name \"%s\".", jsString(raw, present))
	}
	if err := ValidateWorkerName(name, "worker name"); err != nil {
		return Manifest{}, err
	}
	entry, ok := m["entry"].(string)
	if !ok || entry == "" {
		return Manifest{}, fmt.Errorf("Worker manifest at \"%s\" is missing an \"entry\" field.", file)
	}
	if _, err := ResolvePathWithin(workerDir, entry, "Worker entry"); err != nil {
		return Manifest{}, err
	}
	manifest := Manifest{Name: name, Entry: entry}
	if raw, present := m["on_failure"]; present {
		policy, ok := raw.(string)
		if !ok || (policy != "continue" && policy != "error") {
			return Manifest{}, fmt.Errorf("Worker manifest at \"%s\" has an invalid \"on_failure\" field.", file)
		}
		manifest.OnFailure = &policy
	}
	if raw, present := m["location"]; present {
		manifest.Location.Present = true
		if raw != nil {
			location, ok := raw.(map[string]any)
			if !ok {
				return Manifest{}, fmt.Errorf("Worker manifest at \"%s\" has an invalid location.", file)
			}
			op, ok := location["operator"].(string)
			if !ok || !validOperator(op) {
				return Manifest{}, fmt.Errorf("Worker manifest at \"%s\" has an invalid location operator.", file)
			}
			val, ok := location["value"].(string)
			if !ok || val == "" || hasTerminalControls(val) {
				return Manifest{}, fmt.Errorf("Worker manifest at \"%s\" has an invalid location value.", file)
			}
			manifest.Location.Value = &Location{Operator: op, Value: val}
		}
	}
	return manifest, nil
}

// jsString is only used to retain Node's validation diagnostics for non-string fields.
func jsString(value any, present bool) string {
	if !present {
		return "undefined"
	}
	if value == nil {
		return "null"
	}
	switch v := value.(type) {
	case string:
		return v
	case map[string]any:
		return "[object Object]"
	case []any:
		parts := make([]string, len(v))
		for i, x := range v {
			if x != nil {
				parts[i] = jsString(x, true)
			}
		}
		return strings.Join(parts, ",")
	default:
		return fmt.Sprint(value)
	}
}

func DiscoverWorkers(projectDir string) ([]LocalWorker, error) {
	root := filepath.Join(projectDir, WorkersDir)
	entries, err := os.ReadDir(root)
	if errors.Is(err, os.ErrNotExist) {
		return []LocalWorker{}, nil
	}
	if err != nil {
		return nil, err
	}
	workers := []LocalWorker{}
	names := map[string]bool{}
	lower := cases.Lower(language.AmericanEnglish)
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		dir := filepath.Join(root, e.Name())
		if !pathExists(filepath.Join(dir, ManifestFile)) {
			continue
		}
		m, err := ReadManifest(dir)
		if err != nil {
			return nil, err
		}
		normalized := lower.String(m.Name)
		if names[normalized] {
			return nil, fmt.Errorf("Duplicate worker name \"%s\" found in this project.", m.Name)
		}
		names[normalized] = true
		workers = append(workers, LocalWorker{Dir: dir, Manifest: m})
	}
	order := collate.New(language.AmericanEnglish)
	sort.SliceStable(workers, func(i, j int) bool {
		return order.CompareString(workers[i].Manifest.Name, workers[j].Manifest.Name) < 0
	})
	return workers, nil
}

func FindWorker(projectDir, name string) (LocalWorker, error) {
	workers, err := DiscoverWorkers(projectDir)
	if err != nil {
		return LocalWorker{}, err
	}
	for _, w := range workers {
		if w.Manifest.Name == name {
			return w, nil
		}
	}
	for _, w := range workers {
		if filepath.Base(w.Dir) == name {
			return w, nil
		}
	}
	names := make([]string, len(workers))
	for i, w := range workers {
		names[i] = w.Manifest.Name
	}
	available := strings.Join(names, ", ")
	if available == "" {
		available = "(none)"
	}
	return LocalWorker{}, fmt.Errorf("No worker named \"%s\" found in this project. Available workers: %s.", name, available)
}

File: src/lib/edge-workers/project.ts
/**
 * Edge-workers project resolution and on-disk layout helpers.
 *
 * Layout (created by `vip edge-workers init`):
 *
 *   edge-workers/
 *     edge-workers.json        <- project descriptor (toolchain type)
 *     package.json
 *     lib/                     <- shared modules
 *     workers/
 *       <name>/
 *         worker.json          <- per-worker manifest
 *         assembly/index.ts    <- entry (toolchain-specific)
 */

import fs from 'node:fs';
import path from 'node:path';

import UserError from '../user-error';
import { parseProjectDescriptor, parseWorkerManifest } from './validation';

import type { DiscoveredWorker, ProjectDescriptor, WorkerManifest } from './types';

export const PROJECT_DESCRIPTOR_FILE = 'edge-workers.json';
export const WORKER_MANIFEST_FILE = 'worker.json';
export const WORKERS_DIR = 'workers';
/** Conventional output directory for compiled artifacts, relative to the project root. */
export const BUILD_DIR = 'build';
/** Conventional subfolder checked when resolving from a site-repo root. */
export const CONVENTIONAL_PROJECT_DIR = 'edge-workers';

function isProjectRoot( dir: string ): boolean {
	return fs.existsSync( path.join( dir, PROJECT_DESCRIPTOR_FILE ) );
}

/**
 * Read a project file as UTF-8, rejecting symlinks before opening so a symlinked
 * descriptor/manifest can't redirect the read outside the project tree.
 */
function readProjectFile( file: string, label: string ): string {
	let stat: fs.Stats;
	try {
		stat = fs.lstatSync( file );
	} catch {
		throw new UserError( `Could not read ${ label } at "${ file }".` );
	}
	if ( stat.isSymbolicLink() ) {
		throw new UserError( `${ label } at "${ file }" must not be a symbolic link.` );
	}
	try {
		return fs.readFileSync( file, 'utf8' );
	} catch {
		throw new UserError( `Could not read ${ label } at "${ file }".` );
	}
}

/**
 * Resolve the edge-workers project directory for a command.
 *
 * Resolution order:
 *   1. `--path` if provided (must contain a project descriptor).
 *   2. Walk up from the current working directory looking for the descriptor.
 *   3. The conventional `./edge-workers` subfolder, if present.
 *   4. Otherwise throw a UserError with guidance.
 */
export function resolveProjectDir(
	opts: { path?: string } = {},
	cwd: string = process.cwd()
): string {
	if ( opts.path !== undefined ) {
		// `--path` passed without a value arrives as a boolean, not a string;
		// guard so we surface a clear error instead of a throw from path.resolve().
		if ( typeof opts.path !== 'string' || opts.path === '' ) {
			throw new UserError( 'The --path flag requires a path to the edge-workers project.' );
		}

		const explicit = path.resolve( cwd, opts.path );
		if ( ! isProjectRoot( explicit ) ) {
			throw new UserError(
				`No edge-workers project found at "${ explicit }" (missing ${ PROJECT_DESCRIPTOR_FILE }).`
			);
		}

		return explicit;
	}

	// Walk up from cwd.
	let current = path.resolve( cwd );

	while ( true ) {
		if ( isProjectRoot( current ) ) {
			return current;
		}

		const parent = path.dirname( current );
		if ( parent === current ) {
			break;
		}
		current = parent;
	}

	// Conventional subfolder fallback.
	const conventional = path.resolve( cwd, CONVENTIONAL_PROJECT_DIR );
	if ( isProjectRoot( conventional ) ) {
		return conventional;
	}

	throw new UserError(
		'No edge-workers project found here. Run `vip edge-workers init` to create one, ' +
			'run the command from inside a project, or pass `--path` to point at one.'
	);
}

export function readProjectDescriptor( projectDir: string ): ProjectDescriptor {
	const file = path.join( projectDir, PROJECT_DESCRIPTOR_FILE );
	const raw = readProjectFile( file, 'project descriptor' );

	let parsed: unknown;
	try {
		parsed = JSON.parse( raw ) as unknown;
	} catch {
		throw new UserError( `Project descriptor at "${ file }" is not valid JSON.` );
	}

	return parseProjectDescriptor( parsed, file );
}

export function writeProjectDescriptor( projectDir: string, descriptor: ProjectDescriptor ): void {
	const file = path.join( projectDir, PROJECT_DESCRIPTOR_FILE );
	fs.mkdirSync( projectDir, { recursive: true } );
	fs.writeFileSync( file, JSON.stringify( descriptor, null, '\t' ) + '\n' );
}

export function readWorkerManifest( workerDir: string ): WorkerManifest {
	const file = path.join( workerDir, WORKER_MANIFEST_FILE );
	const raw = readProjectFile( file, 'worker manifest' );

	let parsed: unknown;
	try {
		parsed = JSON.parse( raw ) as unknown;
	} catch {
		throw new UserError( `Worker manifest at "${ file }" is not valid JSON.` );
	}

	return parseWorkerManifest( parsed, file );
}

export function writeWorkerManifest( workerDir: string, manifest: WorkerManifest ): void {
	const file = path.join( workerDir, WORKER_MANIFEST_FILE );
	fs.mkdirSync( workerDir, { recursive: true } );
	fs.writeFileSync( file, JSON.stringify( manifest, null, '\t' ) + '\n' );
}

/** Discover all workers in a project by scanning each `workers/<name>/worker.json`. */
export function discoverWorkers( projectDir: string ): DiscoveredWorker[] {
	const workersRoot = path.join( projectDir, WORKERS_DIR );
	if ( ! fs.existsSync( workersRoot ) ) {
		return [];
	}

	const entries = fs.readdirSync( workersRoot, { withFileTypes: true } );
	const workers: DiscoveredWorker[] = [];
	const workersByName = new Map< string, DiscoveredWorker >();
	for ( const entry of entries ) {
		if ( ! entry.isDirectory() ) {
			continue;
		}

		const dir = path.join( workersRoot, entry.name );
		if ( ! fs.existsSync( path.join( dir, WORKER_MANIFEST_FILE ) ) ) {
			continue;
		}

		const worker = { dir, manifest: readWorkerManifest( dir ) };
		const normalizedName = worker.manifest.name.toLocaleLowerCase( 'en-US' );
		if ( workersByName.has( normalizedName ) ) {
			throw new UserError(
				`Duplicate worker name "${ worker.manifest.name }" found in this project.`
			);
		}
		workersByName.set( normalizedName, worker );
		workers.push( worker );
	}

	return workers.sort( ( left, right ) => left.manifest.name.localeCompare( right.manifest.name ) );
}

/**
 * Find a single worker by name (the manifest `name`, falling back to the
 * directory name for convenience).
 */
export function findWorker( projectDir: string, name: string ): DiscoveredWorker {
	const workers = discoverWorkers( projectDir );
	const match =
		workers.find( worker => worker.manifest.name === name ) ??
		workers.find( worker => path.basename( worker.dir ) === name );

	if ( ! match ) {
		const available = workers.map( worker => worker.manifest.name ).join( ', ' ) || '(none)';
		throw new UserError(
			`No worker named "${ name }" found in this project. Available workers: ${ available }.`
		);
	}

	return match;
}
